File: homework0/tom_nguyen/q0_ZeroSum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

// Uses a hash map to track seen values and find complement pairs in O(n)
int zeroSumPairs(const vector<int>& arr) {
    unordered_map<int, int> count;
    for (int num : arr) {
        ++count[num];
    }
    int pairs = 0;
    for (const auto& p : count) {
        int num = p.first;
        int cnt = p.second;
        if (num == 0) {
            pairs += cnt / 2;
            count[0] = cnt % 2;
            continue;
        } 
        if (!count.contains(-num)) {
            continue;
        }
        int match = min(cnt, count[-num]);
        pairs += match;
        count[num] -= match;
        count[-num] -= match;
    }
    return pairs;
}

// Uses a hash map to track seen values and find ALL complement pairs in O(n)
int zeroSumPairsReuse(const vector<int>& arr) {
    unordered_map<int, int> count;
    for (int num : arr) {
        ++count[num];
    }
    int pairs = 0;
    for (auto& p : count) {
        int num = p.first;
        int cnt = p.second;
        if (num == 0) {
            pairs += (cnt * (cnt - 1)) / 2;
            count[0] = 0;
            continue;
        }
        if (num < 0 || !count.contains(-num)) {
            continue;
        }
        pairs += cnt * count[-num];
        count[num] = 0;
        count[-num] = 0;
    }
    return pairs;
}

int main() {
    cout << "Part 1:" << endl;
    vector<int> t1 = {1, 10, 8, 3, 2, 5, 7, 2, -2, -1};
    cout << zeroSumPairs(t1) << endl;

    vector<int> t2 = {1, 10, 8, -2, 2, 5, 7, 2, -2, -1};
    cout << zeroSumPairs(t2) << endl;

    vector<int> t3 = {4, 3, 3, 5, 7, 0, 2, 3, 8, 6};
    cout << zeroSumPairs(t3) << endl;

    vector<int> t4 = {4, 3, 3, 5, 7, 0, 2, 3, 8, 0};
    cout << zeroSumPairs(t4) << endl;

    cout << "\nFollow-up:" << endl;
    vector<int> t5 = {1, 10, 8, 3, 2, 5, 7, 2, -2, -1};
    cout << zeroSumPairsReuse(t5) << endl;

    vector<int> t6 = {1, 10, 8, -2, 2, 5, 7, 2, -2, -1};
    cout << zeroSumPairsReuse(t6) << endl;

    vector<int> t7 = {4, 3, 3, 5, 7, 0, 2, 3, 8, 6};
    cout << zeroSumPairsReuse(t7) << endl;

    vector<int> t8 = {4, 3, 3, 5, 7, 0, 2, 3, 8, 0};
    cout << zeroSumPairsReuse(t8) << endl;

    return 0;
}

// Time spent: 30 minutes

File: rishav_acharya/q0_ZeroSum.py
"""
Q0: Zero Sum Pairs
Given an array of integers, find the number of pairs that sum to zero.
Part 1: Each element can be used at most once
Part 2: Elements can be reused in different pairs (but pairs must use different indices)
"""

def zero_sum_once(arr):
    """
    Find pairs that sum to zero where each element is used at most once.
    
    Algorithm:
    - Use a dictionary to track which numbers we've encountered
    - For each number, check if its negative exists in our tracker
    - If found, we have a pair; remove the complement to prevent reuse
    - Otherwise, add current number to tracker
    """
    tracker = {}
    pair_count = 0
    
    for num in arr:
        complement = -num
        
        # Check if complement exists and hasn't been used
        if complement in tracker and tracker[complement] > 0:
            pair_count += 1
            tracker[complement] -= 1  # Mark as used
        else:
            # Add current number to tracker
            tracker[num] = tracker.get(num, 0) + 1
    
    return pair_count

# Time Complexity: O(n) - single pass through array
# Space Complexity: O(n) - dictionary storage


def zero_sum_unlimited(arr):
    """
    Find all possible pairs that sum to zero (elements can be in multiple pairs).
    
    Algorithm:
    - Maintain frequency count of all elements seen so far
    - For each element, add the frequency of its complement (all previous complements form pairs)
    - Update frequency of current element
    """
    frequency = {}
    total_pairs = 0
    
    for num in arr:
        complement = -num
        
        # Add count of how many times complement appeared before
        if complement in frequency:
            total_pairs += frequency[complement]
        
        # Update frequency of current number
        frequency[num] = frequency.get(num, 0) + 1
    
    return total_pairs

# Time Complexity: O(n) - single traversal
# Space Complexity: O(n) - frequency map storage


if __name__ == "__main__":
    test1 = [1, 10, 8, 3, 2, 5, 7, 2, -2, -1]
    test2 = [1, 10, 8, -2, 2, 5, 7, 2, -2, -1]
    test3 = [4, 3, 3, 5, 7, 0, 2, 3, 8, 6]
    test4 = [4, 3, 3, 5, 7, 0, 2, 3, 8, 0]
    
    print("Test with at most once:", zero_sum_once(test4))
    print("Test with unlimited reuse:", zero_sum_unlimited(test4))

# Time spent: 40 minutes
